### PR TITLE
[SDL 0238] Scripts for `Keyboard Enhancements` feature

### PR DESCRIPTION
**Pull request by GetmanetsIrina**
ATF Test Scripts to check [#2970](https://github.com/smartdevicelink/sdl_core/issues/2970)

This PR is **[ready]** for review.

### Summary
New scripts for the `Keyboard Enhancements` feature

### ATF version
[feature/7.1.0_version_bump](https://github.com/smartdevicelink/sdl_atf/tree/feature/7.1.0_version_bump)

### Changelog
 - Added new scripts for the feature
 - Added new test set for the feature

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

**Comment by KhrystynaDubovyk** [comment]
@ashwink11, please review this PR.

**Comment by ashwink11** [comment]
@KhrystynaDubovyk changes approved from the Ford side.

**Comment by KhrystynaDubovyk** [comment]
@JackLivio This PR is ready for Livio ready. Thank you!

**Comment by ashwink11** [comment]
@jordynmackool  this PR is ready for review. Thanks!

File: test_scripts/API/KeyboardEnhancements/SetGlobalProperties/006_App_defines_invalid_value_in_customizeKeys.lua [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+--  - Respond with INVALID_DATA, success:false to App
+----------------------------------------------------------------------------------------------------
+--[[ Required Shared libraries ]]
+local common = require('test_scripts/API/KeyboardEnhancements/common')
+
+--[[ Local Variables ]]
+local dispCaps = common.getDispCaps()
+dispCaps.systemCapability.displayCapabilities[1].windowCapabilities[1].keyboardCapabilities = {
+  supportedKeyboardLayouts = { "NUMERIC" },
+  configurableKeys = { { keyboardLayout = "NUMERIC", numConfigurableKeys = 11 } }
```
Comment (jacobkeeler) [inline test_scripts/API/KeyboardEnhancements/SetGlobalProperties/006_App_defines_invalid_value_in_customizeKeys.lua:23]:
```suggestion
  configurableKeys = { { keyboardLayout = "NUMERIC", numConfigurableKeys = 10 } }
```

File: test_scripts/API/KeyboardEnhancements/SetGlobalProperties/008_App_defines_not_supported_special_character_in_customizeKeys.lua [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+----------------------------------------------------------------------------------------------------
+--[[ Required Shared libraries ]]
+local common = require('test_scripts/API/KeyboardEnhancements/common')
+
+--[[ Local Functions ]]
+local function sendSetGP()
+  local sgpParams = {
+    keyboardProperties = {
+      keyboardLayout = "NUMERIC",
+      customizeKeys = { "^" } -- not supported special character
```
Comment (jacobkeeler) [inline test_scripts/API/KeyboardEnhancements/SetGlobalProperties/008_App_defines_not_supported_special_character_in_customizeKeys.lua:27]:
```suggestion
      customizeKeys = { "^" } -- special character that is not supported by HMI
```

**Review by jacobkeeler — CHANGES_REQUESTED** [review]

File: test_scripts/API/KeyboardEnhancements/Capabilities/001_HMI_provides_valid_values_for_all_parameters_in_KeyboardCapabilities.lua [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+local common = require('test_scripts/API/KeyboardEnhancements/common')
+
+--[[ Scenario ]]
+common.Title("Preconditions")
+common.Step("Clean environment", common.preconditions)
+common.Step("Start SDL, HMI, connect Mobile, start Session", common.start)
+common.Step("Register App", common.registerApp)
+
+common.Title("Test")
+common.Step("HMI sends OnSCU", common.sendOnSCU)
```
[the remark this answers is not in the document]
Comment (GetmanetsIrina) [reply test_scripts/API/KeyboardEnhancements/Capabilities/001_HMI_provides_valid_values_for_all_parameters_in_KeyboardCapabilities.lua]: @jacobkeeler, fixed in 0239770 .

File: test_scripts/API/KeyboardEnhancements/SetGlobalProperties/005_App_defines_customizeKeys_edge_cases.lua [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+--  - Proceed with request successfully
+----------------------------------------------------------------------------------------------------
+--[[ Required Shared libraries ]]
+local common = require('test_scripts/API/KeyboardEnhancements/common')
+
+--[[ Local Variables ]]
+local dispCaps = common.getDispCaps()
+dispCaps.systemCapability.displayCapabilities[1].windowCapabilities[1].keyboardCapabilities = {
+  supportedKeyboardLayouts = { "NUMERIC" },
+  configurableKeys = { { keyboardLayout = "NUMERIC", numConfigurableKeys = 11 } }
```
[the remark this answers is not in the document]
Comment (GetmanetsIrina) [reply test_scripts/API/KeyboardEnhancements/SetGlobalProperties/005_App_defines_customizeKeys_edge_cases.lua]: @jacobkeeler, fixed in 0239770

File: test_scripts/API/KeyboardEnhancements/SetGlobalProperties/006_App_defines_invalid_value_in_customizeKeys.lua (hunk 1 above)
Comment (GetmanetsIrina) [reply test_scripts/API/KeyboardEnhancements/SetGlobalProperties/006_App_defines_invalid_value_in_customizeKeys.lua]: @jacobkeeler, fixed in 0239770

File: test_scripts/API/KeyboardEnhancements/SetGlobalProperties/008_App_defines_not_supported_special_character_in_customizeKeys.lua (hunk 2 above)
Comment (GetmanetsIrina) [reply test_scripts/API/KeyboardEnhancements/SetGlobalProperties/008_App_defines_not_supported_special_character_in_customizeKeys.lua]: @jacobkeeler, updated in 0239770

File: test_scripts/API/KeyboardEnhancements/common.lua [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+
+--[[ General configuration parameters ]]
+runner.testSettings.isSelfIncluded = false
+config.defaultProtocolVersion = 2
+
+--[[ Local Variables ]]
+local m = {}
+
+--[[ Common Proxy Functions ]]
+do
```
[the remark this answers is not in the document]
Comment (GetmanetsIrina) [reply test_scripts/API/KeyboardEnhancements/common.lua]: This `do` block was only for more convenient code reading (the variable list minimizing in a text editor). It has been removed in 0239770

File: test_scripts/API/KeyboardEnhancements/Capabilities/006_SDL_populates_numConfigurableKeys_with_zeros.lua [hunk 6]
```diff
@@ -0,0 +1,108 @@
+----------------------------------------------------------------------------------------------------
```
[the remark this answers is not in the document]
Comment (GetmanetsIrina) [reply test_scripts/API/KeyboardEnhancements/Capabilities/006_SDL_populates_numConfigurableKeys_with_zeros.lua:1]: The script will be removed after smartdevicelink/sdl_evolution#1117 is accepted.

**Comment by GetmanetsIrina** [comment]
@jacobkeeler, all comments are addressed.

**Comment by GetmanetsIrina** [comment]
d2ad6d0 fixes a typo in the function name.

**Comment by dboltovskyi** [comment]
@jacobkeeler Please notice all updates related to revision are done and PR is ready for review. Thank you!

**Review by jacobkeeler — CHANGES_REQUESTED** [review]

File: test_scripts/API/KeyboardEnhancements/common.lua [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+  m.getHMIAppId = actions.getHMIAppId
+  m.cloneTable = utils.cloneTable
+  m.connectMobile = actions.mobile.connect
+  m.wait = utils.wait
+  m.json = actions.json
+  m.getPolicyAppId = actions.app.getPolicyAppId
+  m.getParams = actions.app.getParams
+  m.spairs = utils.spairs
+  m.isTableEqual = utils.isTableEqual
+  m.tableToString = utils.tableToString
```
[the remark this answers is not in the document]
Comment (dboltovskyi) [reply test_scripts/API/KeyboardEnhancements/common.lua]: Done in 6aa7b8b

**Review by jacobkeeler — APPROVED** [review]

**Pull request merged**